### PR TITLE
feat: implement static importlib support and file deduplication

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

This PR implements comprehensive support for static `importlib.import_module()` calls with literal string arguments, enabling Cribo to bundle modules imported dynamically at runtime. Additionally, it includes foundational work for file-based deduplication as outlined in the implementation proposal.

### 1. Static importlib Import Discovery
- Added detection of `importlib.import_module("literal")` calls during the import discovery phase
- Tracks these as regular imports so they participate in bundling
- Handles both direct calls and calls via imported `import_module` function

### 2. Import Transformation & Rewriting
- Transforms `importlib.import_module("module")` calls to direct module references
- Handles variable assignments: `mod = importlib.import_module("module")` → `mod = bundled_module`
- Supports attribute access on importlib-imported modules: `mod.function()` works correctly
- Maintains namespace objects for inlined modules to preserve module-like behavior

### 3. Module Names with Invalid Python Identifiers
- Supports modules with names that aren't valid Python identifiers (e.g., "data-processor", "2024-config")
- These modules are automatically wrapped since they can only be imported via importlib
- Preserves functionality while ensuring valid Python output

### 4. Tree-shaking Improvements
- Enhanced graph builder to handle `assert` statements
- Tracks variable reads and attribute accesses within assert expressions
- Ensures symbols used in assertions are preserved during tree-shaking

### 5. Import Deduplication & Namespace Management
- Proper handling of types import for namespace objects
- Fixed import hoisting to capture all necessary imports after transformations
- Extracted common module access logic to reduce code duplication

## Test Coverage

Added comprehensive test fixtures:
- `importlib_static/` - Tests various importlib patterns with invalid identifiers
- `importlib_edge_cases/` - Edge cases for dynamic imports
- `file_deduplication_basic/` - Multiple import forms of the same module
- `import_shadowing/` - Complex import shadowing scenarios
- `stickytape_script_with_dynamic_import/` - Basic importlib functionality

### Changes by Component

**Import Discovery (`visitors/import_discovery.rs`)**
- Added `ImportlibStatic` variant to `ImportType` enum
- Implemented detection of static importlib calls in AST visitor
- Tracks both `importlib.import_module()` and `from importlib import import_module` patterns

**Code Generator (`code_generator.rs`)**
- Added `RecursiveImportTransformer` to handle importlib transformations
- Tracks importlib-assigned variables for attribute access rewriting
- Implements `create_module_access_expr()` for unified module access logic
- Fixed import hoisting timing to include all discovered imports

**Graph Builder (`graph_builder.rs`)**
- Added `process_assert_stmt()` for assert statement analysis
- Enhanced expression collection to track assertions properly
- Ensures accurate dependency tracking for tree-shaking

**Orchestrator (`orchestrator.rs`)**
- Enhanced import extraction to include importlib calls
- Improved module discovery for dynamic imports
- Better handling of invalid module identifiers

**Resolver (`resolver.rs`)**
- Added `is_valid_python_identifier()` validation
- Enhanced path resolution for importlib modules
- Improved error messages for module resolution

## Migration Notes

- Tests previously marked as `xfail_` that now pass have been renamed:
  - `xfail_stickytape_script_with_dynamic_import` → `stickytape_script_with_dynamic_import`
  - `xfail_file_deduplication_basic` → `file_deduplication_basic`
  - `xfail_importlib_edge_cases` → `importlib_edge_cases`

## Future Work

This PR lays the groundwork for complete file-based deduplication as described in `docs/file-deduplication-implementation-proposal.md`. The remaining xfail tests indicate areas for future enhancement:
- `xfail_file_deduplication_symlinks` - Symlink handling
- `xfail_importlib_deduplication` - Complex deduplication scenarios

## Test Plan

- [x] All existing tests pass
- [x] New test fixtures demonstrate correct behavior
- [x] Bundled output executes correctly
- [x] Tree-shaking preserves all used symbols
- [x] Invalid identifiers are handled properly

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by tinovyatkin** [comment]
@coderabbitai Thanks for catching the duplicate ImportlibStatic imports issue!

✅ Fixed in 6e10ae4 

The issue was that `walk_stmt` was being called twice for assignment statements - once explicitly in the match arm and then again at the end of `visit_stmt`. Added the missing `return` statement to prevent duplicate traversal.

The deduplication logic is no longer needed and has been removed from the test.

File: crates/cribo/src/cribo_graph.rs [hunk 1]
```diff
@@ -712,24 +722,88 @@
...
+                if existing_canonical == &canonical_path {
+                    return existing_id; // Same import name, same file - reuse
+                } else {
+                    // Error: same import name but different files
+                    // This shouldn't happen with proper PYTHONPATH management
+                    log::error!(
+                        "Import name '{name}' refers to different files: {existing_canonical:?} \
+                         and {canonical_path:?}"
+                    );
+                }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/cribo_graph.rs]: ✅ Addressed in 69bbd95. Thanks\!

File: crates/cribo/src/cribo_graph.rs [hunk 2]
```diff
@@ -1460,6 +1536,38 @@
...
+    /// Get the canonical path for a module
+    pub fn get_canonical_path(&self, module_id: ModuleId) -> Option<&PathBuf> {
+        self.module_canonical_paths.get(&module_id)
+    }
+
+    /// Get the primary module for a given file path
+    pub fn get_primary_module_for_file(&self, path: &Path) -> Option<(String, ModuleId)> {
+        let canonical = path.canonicalize().unwrap_or_else(|_| path.to_path_buf());
+        self.file_primary_module.get(&canonical).cloned()
+    }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/cribo_graph.rs:1591]: ✅ Addressed in eda2425. Thanks\!

File: crates/cribo/src/code_generator.rs [hunk 3]
```diff
@@ -3129,8 +3383,49 @@ impl HybridStaticBundler {
...
+                            _ => {
+                                log::debug!(
+                                    "  Statement {}: {:?}",
+                                    i,
+                                    std::mem::discriminant(stmt)
+                                );
+                            }
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/code_generator.rs]: ✅ Addressed in 82ea097. Thanks\!

File: crates/cribo/src/code_generator.rs [hunk 4]
```diff
@@ -1795,6 +1993,65 @@
...
+        match stmt {
+            Stmt::Expr(expr_stmt) => self.expr_uses_importlib(&expr_stmt.value),
+            Stmt::Assign(assign) => self.expr_uses_importlib(&assign.value),
+            Stmt::FunctionDef(func_def) => {
+                func_def.body.iter().any(|s| self.stmt_uses_importlib(s))
+            }
+            Stmt::ClassDef(class_def) => class_def.body.iter().any(|s| self.stmt_uses_importlib(s)),
+            _ => false,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/code_generator.rs]: ✅ Addressed in d0557fd. Thanks\!

Expanded the `stmt_uses_importlib` function to comprehensively check all statement types including:
- Control flow statements (If, While, For, With, Try)
- Expression statements (Assert, Return, Raise, Delete)
- Pattern matching (Match) and type aliases (TypeAlias)

Also enhanced `expr_uses_importlib` to handle all expression types including comprehensions, lambda expressions, boolean operations, and more.

All tests pass and clippy is clean.

File: crates/cribo/src/orchestrator.rs [hunk 5]
```diff
@@ -865,8 +889,16 @@
...
+            // Handle ImportlibStatic imports
+            if matches!(
+                import.import_type,
+                crate::visitors::ImportType::ImportlibStatic
+            ) {
+                if let Some(ref module_name) = import.module_name {
+                    debug!("Found ImportlibStatic import: {module_name}");
+                    imports_set.insert(module_name.clone());
+                }
+            } else if import.level > 0 {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs]: ✅ Addressed in 0bc2b72. Thanks\!

Extracted the ImportlibStatic handling logic into a dedicated helper method `process_importlib_static_import`. This eliminates the code duplication between `extract_all_imports` and `extract_all_imports_with_context` methods.

The helper method encapsulates the common logic for processing ImportlibStatic imports and updating the imports set, following DRY principles.

All tests pass and clippy is clean.

File: crates/cribo/src/code_generator.rs [hunk 6]
```diff
@@ -12745,4 +13231,149 @@ impl HybridStaticBundler {
...
+                symbol_renames.get(module_name),
+            )
+        } else {
+            // This module wasn't bundled - shouldn't happen for static imports
+            // Return None literal as a fallback
+            Expr::NoneLiteral(ExprNoneLiteral {
+                node_index: AtomicNodeIndex::dummy(),
+                range: TextRange::default(),
+            })
+        }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/code_generator.rs]: ✅ Addressed in c16a942

Added warning log for unbundled modules in static imports. Now when a module referenced in a static import is not bundled, it logs a warning to help catch configuration issues early.

Thanks for the suggestion\!

File: crates/cribo/src/orchestrator.rs [hunk 7]
```diff
@@ -772,13 +773,36 @@ impl BundleOrchestrator {
...
+        debug!(
+            "ImportDiscoveryVisitor found {} imports",
+            discovered_imports.len()
+        );
+        for (i, import) in discovered_imports.iter().enumerate() {
+            debug!(
+                "Import {}: type={:?}, module={:?}",
+                i, import.import_type, import.module_name
+            );
+        }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs]: ✅ Addressed in 971f06b

Changed detailed per-import logging from debug to trace level:
- Summary count still logged at debug level
- Detailed enumeration only happens when trace logging is enabled
- This reduces log noise in production environments with large codebases

Thanks for the suggestion\!

File: crates/cribo/src/orchestrator.rs [hunk 8]
```diff
@@ -1289,38 +1321,73 @@
...
+                self.add_to_discovery_queue_if_new(import, import_path, params);
+            } else {
+                // Try normal resolution in case it's a valid Python identifier
+                match params.resolver.classify_import(import) {
+                    ImportType::FirstParty => {
+                        if let Ok(Some(import_path)) = params.resolver.resolve_module_path(import) {
+                            debug!(
+                                "Resolved ImportlibStatic '{import}' via normal resolution to \
+                                 path: {import_path:?}"
+                            );
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs]: ✅ Addressed in a47c4aa

Simplified the debug logging by removing the redundant "via normal resolution" phrase. The resolution path is already implied by the code context.

Thanks for the suggestion\!

File: crates/cribo/src/orchestrator.rs [hunk 9]
```diff
@@ -1289,38 +1343,117 @@ impl BundleOrchestrator {
...
+                } else if !is_in_error_handler {
+                    return Err(anyhow!(
+                        "Failed to resolve ImportlibStatic module '{}' (resolved from '{}'). \
+                         \nThis import would fail at runtime with: ModuleNotFoundError: No module \
+                         named '{}'",
+                        resolved_name,
+                        import,
+                        resolved_name
+                    ));
+                }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs]: ✅ Addressed in 08a5eb2. Thanks\!

Modified `resolve_importlib_static_with_context` to return both the resolved module name and path as a tuple, eliminating the need to duplicate the relative import resolution logic in the orchestrator. This also fixed the issue where relative imports like `.submodule` were being added as module names.

File: crates/cribo/src/resolver.rs [hunk 10]
```diff
@@ -452,20 +466,114 @@ impl ModuleResolver {
...
+                    }
+                }
+                module_path = module_path.join(component);
+            }
+
+            // Try as a package directory with __init__.py
+            let init_path = module_path.join("__init__.py");
+            if init_path.is_file() {
+                debug!("Found ImportlibStatic package at: {init_path:?}");
+                return Ok(Some(init_path));
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/resolver.rs]: ✅ Addressed in 6e761e4. Thanks\!

Added path canonicalization to all return statements in `resolve_importlib_static_with_context` to ensure consistency with other resolution methods. The method now properly canonicalizes paths before returning them as part of the tuple.
